File: python/medium/trade.py
import math

prices_1 = [10, 1, 2, 4, 11]
prices_2 = [27,53,7,25,33,2,32,47,43]

def trade_slow(l):
    max_income = 0
    best_prices = [0,0]
    for price in l:
        for item in l:
            candidate = price - item
            if candidate > max_income:
                max_income = candidate
                best_prices[0] = item
                best_prices[1] = price
    return max_income, best_prices
print(trade_slow(prices_1))    # O(n^2)

def trade(l):
    if len(l) == 1:
        return 0
    mid = math.floor(len(l) / 2)
    former = l[0:mid]
    latter = l[mid:]
    case_3 = max(latter) - min(former)
    return max(trade(former), trade(latter), case_3)
print(trade(prices_1))  # O(nlogn)

def trade_dp(p):
    if not p:
        return None
    min_prices = [p[0]]
    sell_day_idx = 0
    buy_day_idx = 0
    best_profit = 0

    for i in range(1, len(p)):
        n = p[i]
        if n < min_prices[i-1]:
            min_prices.append(n)
        else:
            min_prices.append(min_prices[i-1])
            
        profit = n - min_prices[i]
        if profit > best_profit:
            sell_day_idx = i
            best_profit = profit
            buy_day_idx = p.index(min_prices[i])
    return (buy_day_idx, sell_day_idx)
print(trade_dp(prices_1))   # O(n)

def trade_kadane(p):
    min_price = p[0]
    sell_day_idx = 0
    buy_day_idx = 0
    best_profit = 0

    for i in range(1, len(p)):
        if p[i] < min_price:
            min_price = p[i]
            buy_day_idx = i
        profit = p[i] - min_price

        if profit > best_profit:
            sell_day_idx = i
            best_profit = profit
    return (buy_day_idx,sell_day_idx)
print(trade_kadane(prices_1))

File: ts/medium/trade.ts
const prices = [10, 1, 2, 4, 11];

function naiveTrade(p: number[]): [number, number, number] {
  let maxProfit = 0;
  let result: [number, number, number] = [0, 0, 0];
  for (const iPrice of p) {
    for (const jPrice of p) {
      const candidate = iPrice - jPrice;
      if (candidate > maxProfit) {
        maxProfit = candidate;
        result = [maxProfit, iPrice, jPrice];
      }
    }
  }
  return result;
}
console.log(naiveTrade(prices));  //  O(N^2)

function findMaxInArr(arr: number []): number {
  let max = arr[0];
  for (const num of arr) {
    if (num > max) {
      max = num;
    }
  }
  return max;
}

function findMinInArr(arr: number []): number {
  let min = arr[0];
  for (const num of arr) {
    if (num < min) {
      min = num;
    }
  }
  return min;
}

function trade(p: number[]): number | number[] {
  if (p.length === 1) return 0;

  const mid = Math.floor(p.length / 2);
  const first = p.slice(0, mid);
  const second = p.slice(mid);
  const case3 = findMaxInArr(second) - findMinInArr(first);
  return findMaxInArr([...trade(first) as number[], ...trade(second) as number[], case3]);
}
console.log(trade(prices));  //  O(NlogN)

function tradeDp(p: number[]): [number, number] {
  let minPrices = [p[0]];
  let sellDayIdx = 0;
  let buyDayIdx = 0;
  let bestProfit = 0;

  for (let i = 0; i < p.length; i++) {
    const n = p[i];
    if (n < minPrices[i - 1]) {
      minPrices.push(n);
    } else {
      minPrices.push(minPrices[i -1]);
    }
    const profit = n - minPrices[i];
    if (profit > bestProfit) {
      bestProfit = profit;
      sellDayIdx = i;
      buyDayIdx = p.indexOf(minPrices[i]);
    }
  }
  return [buyDayIdx, sellDayIdx];
}
console.log(tradeDp(prices));

function tradeKadane(p: number[]): [number, number] {
  let minPrice = p[0];
  let sellDayIdx = 0;
  let buyDayIdx = 0;
  let bestProfit = 0;

  for (let i = 0; i < p.length; i++) {
    const n = p[i];
    if (n < minPrice) {
      minPrice = n;
      buyDayIdx = i;
    }
    const profit = n - minPrice;
    if (profit > bestProfit) {
      bestProfit = profit;
      sellDayIdx = i;
    }
  }
  return [buyDayIdx, sellDayIdx];
}
console.log(tradeKadane(prices));